node/src/components/consensus/protocols/highway.rs

```diff
@@ -538,27 +442,41 @@ where
                     None => false,
                 };
 
+                let now = Timestamp::now();
                 match pvv.timestamp() {
-                    Some(timestamp) if timestamp > Timestamp::now() => {
+                    Some(timestamp)
+                        if timestamp > now + self.synchronizer.pending_vertex_timeout() =>
+                    {
+                        trace!("received a vertex with a timestamp far in the future; dropping");
+                        vec![]
+                    }
+                    Some(timestamp) if timestamp > now && is_faulty => {
```

Comment:
Why does it matter that a message is from the future? If a validator is faulty but the message is from "now", we would add it?

Reply:
Apparently we wouldn't: https://github.com/CasperLabs/casper-node/pull/785/files#diff-a2c951f6efd58f4890ffbf04a54d3527c0ebbe0d985089ed8e5df0b6d2c61688R475-R478 

If that's true then I would remove the `timestamp > now` condition as it makes it look like it matters.

Reply:
I didn't actually change that part of the logic in this PR.
I think the reason may be that we had always added all incoming vertices, and then added logic for handling vertices with future timestamp. When we did that, we decided that faulty validators don't deserve putting vertices with future timestamps into the queue.

But you're right, it would make sense to me to just always reject vertices from faulty validators, except if they are dependencies of other queue entries. I'll do that in this PR, if that's okay.